cmd/autobumper/main.go

```diff
@@ -2,6 +2,286 @@
+	return nil
+}
+
+func tagFromName(name string) string {
+	parts := strings.Split(name, ":")
+	if len(parts) < 2 {
+		return ""
+	}
+	return parts[1]
+}
+
+func componentFromName(name string) string {
+	s := strings.Split(strings.Split(name, ":")[0], "/")
+	return s[len(s)-1]
+}
+
+func formatTagDate(d string) string {
+	if len(d) != 8 {
+		return d
+	}
+	// &#x2011; = U+2011 NON-BREAKING HYPHEN, to prevent line wraps.
+	return fmt.Sprintf("%s&#x2011;%s&#x2011;%s", d[0:4], d[4:6], d[6:8])
+}
+
+func generateSummary(name, repo, prefix string, summarise bool, images map[string]string) string {
+	type delta struct {
+		oldCommit string
+		newCommit string
+		oldDate   string
+		newDate   string
+		variant   string
+		component string
+	}
+	versions := map[string][]delta{}
+	for image, newTag := range images {
+		if !strings.HasPrefix(image, prefix) {
+			continue
+		}
+		if strings.HasSuffix(image, ":"+newTag) {
+			continue
+		}
+		oldDate, oldCommit, oldVariant := bumper.DeconstructTag(tagFromName(image))
+		newDate, newCommit, _ := bumper.DeconstructTag(newTag)
+		k := oldCommit + ":" + newCommit
+		d := delta{
+			oldCommit: oldCommit,
+			newCommit: newCommit,
+			oldDate:   oldDate,
+			newDate:   newDate,
+			variant:   oldVariant,
+			component: componentFromName(image),
+		}
+		versions[k] = append(versions[k], d)
+	}
+
+	switch {
+	case len(versions) == 0:
+		return fmt.Sprintf("No %s changes.", name)
+	case len(versions) == 1 && summarise:
+		for k, v := range versions {
+			s := strings.Split(k, ":")
+			return fmt.Sprintf("%s changes: %s/compare/%s...%s (%s → %s)", name, repo, s[0], s[1], formatTagDate(v[0].oldDate), formatTagDate(v[0].newDate))
+		}
+	default:
+		changes := make([]string, 0, len(versions))
+		for k, v := range versions {
+			s := strings.Split(k, ":")
+			names := make([]string, 0, len(v))
+			for _, d := range v {
+				names = append(names, d.component+d.variant)
+			}
+			sort.Strings(names)
+			changes = append(changes, fmt.Sprintf("%s/compare/%s...%s | %s&nbsp;&#x2192;&nbsp;%s | %s",
+				repo, s[0], s[1], formatTagDate(v[0].oldDate), formatTagDate(v[0].newDate), strings.Join(names, ", ")))
+		}
+		sort.Slice(changes, func(i, j int) bool { return strings.Split(changes[i], "|")[1] < strings.Split(changes[j], "|")[1] })
+		return fmt.Sprintf("Multiple distinct %s changes:\n\nCommits | Dates | Images\n--- | --- | ---\n%s\n", name, strings.Join(changes, "\n"))
+	}
+	panic("unreachable!")
+}
+
+func generatePRBody(images map[string]string) string {
+	prowSummary := generateSummary("Prow", prowRepo, prowPrefix, true, images)
+	testImagesSummary := generateSummary("test-image", testImageRepo, testImagePrefix, false, images)
+	body := prowSummary + "\n\n" + testImagesSummary + "\n"
+	return body
+}
+
 func main() {
+	o := parseOptions()
+	if err := validateOptions(o); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+
+	sa := &secret.Agent{}
+	if err := sa.Start([]string{o.githubToken}); err != nil {
```

Comment:
@stevekuznetsov  We really need to expose [this](https://github.com/kubernetes/test-infra/blob/master/prow/config/secret/agent.go#L100). For cases like this, we don't need to `Start` the secretAgent...

Reply:
Technically if we rotate the secret while this tool is slowly working ... this will reload it :)